File: src/main/java/com/example/backend/config/MongoConfig.java
package com.example.backend.config;

import io.mongock.runner.springboot.EnableMongock;
import org.springframework.context.annotation.Configuration;
import org.springframework.data.mongodb.repository.config.EnableMongoRepositories;

//
//import com.mongodb.reactivestreams.client.MongoClient;
//import com.mongodb.reactivestreams.client.MongoClients;
//import org.springframework.context.annotation.Bean;
//import org.springframework.context.annotation.Configuration;
//import org.springframework.data.mongodb.config.AbstractReactiveMongoConfiguration;
//import org.springframework.data.mongodb.core.ReactiveMongoTemplate;
//import org.springframework.data.mongodb.repository.config.EnableMongoRepositories;
//
@Configuration
//@EnableMongoRepositories
@EnableMongock
public class MongoConfig
//        extends AbstractReactiveMongoConfiguration
{
//
//    @Override
//    protected String getDatabaseName() {
//        return "tarot";
//    }
//
////    @Override
////    @Bean
////    public MongoClient reactiveMongoClient() {
////        return MongoClients.create("mongodb://root:mongopw@localhost:27017");
////    }
//
//    @Bean
//    public ReactiveMongoTemplate reactiveMongoTemplate() {
//        return new ReactiveMongoTemplate(reactiveMongoClient(), getDatabaseName());
//    }
//
////    @Bean
////    @ConditionalOnProperty(prefix = "job.autorun", name = "enabled", havingValue = "true", matchIfMissing = true)
////    public CommandLineRunner loadData(UserRepository repository) {
////        return (args) -> {
////            // save a couple of users
////            var users = Flux.just(
////                    new User("Gökhan", "1", 1),
////                    new User("Betül", "2", 12),
////                    new User("Zühtü", "3", 13)
////            );
////            repository.saveAll(users).subscribe();
////        };
////    }
}

File: src/main/kotlin/com/example/backend/config/MongoConfig.kt
package com.example.backend.config

import com.example.backend.repos.JobRepository
import com.example.backend.repos.TarotCardRepository
import com.mongodb.reactivestreams.client.MongoClient
import com.mongodb.reactivestreams.client.MongoClients
import io.mongock.api.config.MongockConfiguration
import io.mongock.driver.mongodb.reactive.driver.MongoReactiveDriver
import org.springframework.beans.factory.annotation.Value
import org.springframework.context.annotation.Bean
import org.springframework.data.mongodb.config.AbstractReactiveMongoConfiguration
import org.springframework.data.mongodb.core.ReactiveMongoTemplate
import org.springframework.data.mongodb.repository.config.EnableReactiveMongoRepositories

@EnableReactiveMongoRepositories(
    basePackageClasses = [
        JobRepository::class,
        TarotCardRepository::class,
    ]
)
//@EnableMongock
class MongoConfig : AbstractReactiveMongoConfiguration() {

    override fun getDatabaseName() = "tarot"

    @Bean
    override fun reactiveMongoClient() = MongoClients.create()

    @Bean
    fun reactiveMongoTemplate(): ReactiveMongoTemplate {
        return ReactiveMongoTemplate(reactiveMongoClient(), databaseName)
    }

    @Bean
    fun connectionDriver(
        @Value("\${spring.data.mongodb.database}") database: String,
        config: MongockConfiguration,
        client: MongoClient,
    ): MongoReactiveDriver {
        val driver = MongoReactiveDriver.withLockStrategy(
            client, database,
            config.lockAcquiredForMillis,
            config.lockQuitTryingAfterMillis,
            config.lockTryFrequencyMillis
        )
        return driver
    }
}